btree/btree.c

```diff
@@ -4075,192 +4223,605 @@ static int  vkvv_space(const struct nd *node)
+{
+	/**
+	 * @brief This function is used to get the size of the key at the given
+	 *        index. This can be achieved by checking the difference in
+	 *        offsets of the current key and the key immediately after it.
+	 */
+	struct vkvv_head *h = vkvv_data(node);
+	if (h->vkvv_level == 0)
+		return vkvv_lnode_rec_key_size(node, idx);
+	else
+		return vkvv_inode_rec_key_size(node, idx);
+}
+
+static uint16_t vkvv_lnode_rec_val_size(const struct nd *node, int idx)
+{
+	struct vkvv_head *h   = vkvv_data(node);
+	struct dir_rec   *rec = (void *)h +sizeof(*h) + h->vkvv_dir_offset;
+
+	return rec[idx+1].val_offset - rec[idx].val_offset;
+}
+
+static uint16_t vkvv_rec_val_size(const struct nd *node, int idx)
+{
+	/**
+	 * @brief This function is used to get the size of the value at the
+	 *        given index. This can be achieved by checking the difference
+	 *        in offsets of the current key and the key immediately after
+	 *        it.
+	 */
+	struct vkvv_head *h = vkvv_data(node);
+	if (h->vkvv_level == 0)
+		return vkvv_lnode_rec_val_size(node,idx);
+	else
+		return vkvv_valsize(node);
+}
+
+static void *vkvv_lnode_key(const struct nd *node, int idx)
+{
+	struct vkvv_head *h   = vkvv_data(node);
+	struct dir_rec   *rec = (void *)h +sizeof(*h) + h->vkvv_dir_offset;
+
+	return ((void*)h + sizeof(*h) + rec[idx].key_offset);
+}
+
+static void *vkvv_inode_key(const struct nd *node, int idx)
+{
+	struct vkvv_head *h = vkvv_data(node);
+	uint16_t         *offset;
+
+	if (idx == 0)
+		return ((void*)h + sizeof(*h));
+	else {
+		offset = vkvv_get_key_offset(node,idx-1);
+		return ((void*)h + sizeof(*h) + *offset);
+	}
+}
+
 static void *vkvv_key(const struct nd *node, int idx)
 {
 	/**
 	 * @brief Return the memory address pointing to key space.
-	 *
-	 * index = given index
-	 * start_addr = node->n_addr + sizeof(node_header)
-	 * new_key = start_addr + dir[index].key_offset
-	 * return new_key
 	 */
+
+	if (vkvv_level(node) == 0)
+		return vkvv_lnode_key(node,idx);
+	else
+		return vkvv_inode_key(node,idx);
+}
+
+static void *vkvv_lnode_val(const struct nd *node, int idx)
+{
+	struct vkvv_head *h    = vkvv_data(node);
+	int               size = 1ULL << h->vkvv_shift;
+	struct dir_rec   *rec  = vkvv_get_dir_addr(node);
+
+	return ((void*)h + size - rec[idx].val_offset);
+}
+
+static void *vkvv_inode_val(const struct nd *node, int idx)
+{
+	struct vkvv_head *h      = vkvv_data(node);
+	uint16_t          size   = 1ULL << h->vkvv_shift;
+	uint16_t          vspace = vkvv_get_vspace();
+
+	return ((void*)h + size - vspace*idx + INT_OFFSET);
 }
 
 static void *vkvv_val(const struct nd *node, int idx)
 {
 	/**
 	 * @brief Return the memory address pointing to value space.
-	 *
-	 * index = given index
-	 * start_addr =  node->n_addr + segaddr_shift(node->n_addr) - 1
-	 * new_val = start_addr - dir[index].val_offset
-	 * return new_val
 	 */
+	if (vkvv_level(node) == 0)
+		return vkvv_lnode_val(node,idx);
+	else
+		return vkvv_inode_val(node,idx);
```

Comment:
for line no:4457 and 4459: s[ace after ','

Reply:
Done